### DISCUSSION TITLE
Install/Uninstall packages in pre/post install/update event script

### DISCUSSION THREAD

**Discussion by Blacksmoke16** [composer/composer#11841] [Q&A / Support] [answered]
As part of a private project, I merged various internal composer libraries into a singular `common` monorepo package for ease of maintenance and such. However, the con of this approach is that some applications this `common` package is installed in, now have additional composer dependencies being installed (due to merging of the dependencies of each internal lib that may not have been installed in each application). This in itself is not really problem, just marginally less efficient.

I had the idea of using the `extra` key in each core application to whitelist what sub-packages are needed from the `common` package. Then, tap into a pre/post update/install hook to either install the extra dependencies required by each whitelisted package, or go back and remove the packages that are not being used.

I was able to get both approaches somewhat working by creating either `InstallOperation` or `UninstallOperation` and executing them. However in both cases they did not seem to take dependencies into account. I.e. installing/removing only the extra packages themselves, but not their dependencies. 

So I guess in short, is there a better way to go about installing/uninstalling packages within an install/update hook that accounts for dependencies? Or alternatively, allow mutating the `requires` list of a specific package before installing/updating? It seems this was discussed a bit as part of https://github.com/composer/composer/discussions/11392, but don't think you'd have access to the `extra` object in that hook?

**Comment by Seldaek** [comment, marked as answer]
I think it'd be much easier to use `replace` to eliminate packages, like what is described as "Composer replace option" on this blog post for polyfills https://php.watch/articles/composer-replace-polyfills - that'd allow you from your root project to get rid of some sub-components you do not want. 

But that sounds to me tricky and dangerous as you may mess things up. I think I'd rather live with a few extra deps installed if I were you.

**Reply by Blacksmoke16** [reply]
Yea, we just ended up accepting we'll have some extra deps hanging around. But the ease of maintenance around having everything centralized is worth the trade-off.
